# rm(list = ls())
# setwd(dirname(rstudioapi::getSourceEditorContext()$path))

library(openxlsx, lib.loc="/home/morril01/R/x86_64-pc-linux-gnu-library/4.0/")
library(ggrepel)
library(tximport, lib.loc = "~/R/x86_64-pc-linux-gnu-library/3.6/")
library(GenomicRanges)

type_counts = 'TPM'

chromlens = rbind.data.frame(c('chr1', 249250621),
                             c('chr2', 243199373),
                             c('chr3',  198022430),
                             c('chr4', 191154276),
                             c('chr5', 180915260),
                             c('chr6', 171115067),
                             c('chr7', 159138663),
                             c('chr8', 146364022),
                             c('chr9', 141213431),
                             c('chr10', 135534747),
                             c('chr11', 135006516),
                             c('chr12', 133851895),
                             c('chr13', 115169878),
                             c('chr14', 107349540),
                             c('chr15', 102531392),
                             c('chr16',  90354753),
                             c('chr17',  81195210),
                             c('chr18',  78077248),
                             c('chr19',  59128983),
                             c('chr20',  63025520),
                             c('chr21',  48129895),
                             c('chr22',  51304566),
                             c('chrX', 155270560),
                             c('chrY',  59373566))
colnames(chromlens) = c('Chrom', 'Length')

# Context:
#-------------
#??Maria needs to check the link between copy number of gene expression for a manuscript on organoids ready for submission (little more information provided)

#??Aim:
#-------------
# Check correlation between absolute copy number and gene expression in 18 samples.
# For a set of genes Maria is interested in,
# or across the genome depending on Maria's decision

# Note:
#-------------
# the code comprises chunks for featureCounts and or kallisto text output that
# we do  not need, but want to keep until the project is finished. 

# Data sets
#-------------

# Sample sheet
#??samples.xlsx: sample sheet with the two sets of sample IDs used: one for CN, the other for gene expression
# samples.csv: text version of samples.xlsx

# CN:
#??organoidAbsolute_SegTable.txt: absolute copy number for 18 samples (columns) and 30 kb bins along autosomes 
#??organoidAbsolute_geneList.txt: table in the long format with genes of interest their coordinates, absolute copy number ('segVal') and sample ('sample')

# gene expression
#??featureCounts, now replaced with kallisto??
# kallisto output in Kallisto2

#??Functions
#-------------

#------------------------------------------------------------------------------
matFromFeatureCounts <- function (sampleTable, directory = ".", ignoreRank = FALSE, ...) {
  #------------------------------------------------------------------------------
  # IN: sampleTable, see DESeqDataSetFromHTSeqCount()
  #	at https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/DESeqDataSet-class
  #	The first column is the sample name, the second column the file name of the count file generated by htseq-count
  
  # read files in:
  l <- lapply(as.character(sampleTable[, 2]), function(fn) read.table(file.path(directory, fn), skip=2))
  # check that all gene names are in the same order:
  if (!all(sapply(l, function(a) all(a$V1 == l[[1]]$V1)))) 
    stop("Gene IDs (first column) differ between files.")
  # Get counts
  tbl <- sapply(l, function(a) a$V7)
  # tidy:
  colnames(tbl) <- sampleTable[, 1]
  rownames(tbl) <- l[[1]]$V1
  rownames(sampleTable) <- sampleTable[, 1]
  return(dds)
}



# Packages
#-------------
library(dplyr)
library(GenomicRanges)

#??Variables
#-------------
projName <- "20191218_ViasM_BJ_orgaBrs"
projDir <- "/mnt/scratcha/fmlab/morril01/Vias/20191218_ViasM_BJ_orgaBrs/GexVsCnGw/"


# Assess correlation between gene expression and copy number.
#-------------

setwd("../Data")

# Read sample sheet in
#-------------
# samples.csv has three columns:
# organoid_name;
# JBLAB-number; # eg JBLAB-19902
#??OV04_number
tmpFn <- "../RnaSeqPip/samplesheet.csv"
splSht <- read.table(tmpFn, sep=",", header=TRUE)
# # add name of fatureCounts file. Mind that is not used anymore, now using Kallisto's TPM.
# #??keeping featureCounts in code though, to compare with kallisto if need be
splSht <- splSht %>% mutate(SampleName=gsub("-","",FSFSampleName)) %>%
  mutate(ftcFn=paste0(SampleName,".featureCounts.gz"))

#??Read CNV data in:
#-------------
# 1: the whole table
tmpFn <- "../Input/organoidAbsolute_SegTable.txt"
absCnDf <- read.table(tmpFn, sep="\t", header=TRUE)

# Make GRanges from first column
#-------------
# Get gene list
#??Get overlap

#??derive data frame to feed makeGRangesFromDataFrame()
tmpl <- strsplit(as.character(absCnDf[,1]),"[:-]")
cnSeg <- do.call(rbind, lapply(tmpl,as.numeric))
colnames(cnSeg) <- c("chr", "start", "end")
# make GRanges, keeping the CN values.
cnSegRg <- makeGRangesFromDataFrame(cbind(cnSeg,absCnDf[,-1]), keep.extra.columns=TRUE)

# Read gene list in ## this is no longer there??
#-------------
tmpFn <- "../RnaSeqPip/organoidAbsolute_geneList.txt"
geneDf <- read.table(tmpFn, sep="\t", header=TRUE)
geneDfRg <- makeGRangesFromDataFrame(geneDf)

# concat featureCount outcome for each sample into a single table:
#-------------
if(FALSE){ #??not using featureConts anymore
  dd <- matFromFeatureCounts(sampleTable = splSht[,c("SampleName","ftcFn")],
                             directory = "samplebam")
}

# read kallisto data in
#-------------
# and get gene-level TPM
#??so first need transcript to gene keys.


#??gene annotation for kallisto
#-------------------------
# data.frame called tx2gene with two columns: 1) transcript ID and 2) gene ID

#??See tximport vignette:
#??Note: if you are using an Ensembl transcriptome, the easiest way to create the tx2gene data.frame is to use the ensembldb packages. The annotation packages can be found by version number, and use the pattern EnsDb.Hsapiens.vXX. The transcripts function can be used with return.type="DataFrame", in order to obtain something like the df object constructed in the code chunk above. See the ensembldb package vignette for more details.

# projDir
setwd("../")
ref_dir <- 'Kallisto/homo_sapiens'
gtf.file <- file.path(ref_dir, "Homo_sapiens.GRCh38.96.gtf.gz")
sqlite_file <- 'Homo_sapiens.GRCh38.96.sqlite'
sqlite_path <- file.path(ref_dir, sqlite_file)

if(!file.exists(sqlite_path)) {
  ## generate the SQLite database file
  ensembldb::ensDbFromGtf(gtf=gtf.file, path = ref_dir, outfile=sqlite_file)
}
EnsDb.Hsapiens.v96 <- ensembldb::EnsDb(sqlite_file)

# Genes, used to annotated the TPM matrix to send to Maria
ag <- ensembldb::genes(EnsDb.Hsapiens.v96, filter=list(AnnotationFilter::GeneBiotypeFilter('protein_coding')), return.type="DataFrame") 
ag

transc <- ensembldb::transcripts(EnsDb.Hsapiens.v96, filter=list(AnnotationFilter::GeneBiotypeFilter('protein_coding')), return.type="DataFrame") 

#??Transcripts:
Tx <- ensembldb::transcripts(EnsDb.Hsapiens.v96,
                             columns = c("tx_id", "gene_id"),
                             return.type="DataFrame")
##head(Tx)

# Write table to file, for reference
# tmpFn <- file.path(ref_dir, "Homo_sapiens.GRCh38.96.txt")
# write.table(Tx, file=tmpFn, sep=",", row.names = FALSE)
rm(Tx)

# read that transcript-gene table in
# (no need here really, but will only write the table above once)
tx2gene <- read.csv(tmpFn)
head(tx2gene)

# Kallisto with abundance.h5
#-------------------------
setwd(projDir)

# have names of files to read in:
files <- file.path("Kallisto2", splSht$JBLAB.number, "abundance.h5")
names(files) <- splSht$SampleName

# collapse transcript counts into gene counts:
#??"We can avoid gene-level summarization by setting txOut=TRUE, giving the
#original transcript level estimates as a list of matrices."
# with txOut = FALSE and passing tx2gene
candidate_kallisto_files <- unlist(sapply(list.files("../Kallisto/", full.names = T), list.files, full.names=T))
files <- base::subset(x = candidate_kallisto_files,
                      subset=grepl('abundance.tsv', candidate_kallisto_files))
example_transcripts=head(read.table(files[1], h=T))
## tx2gene is missing. I make it up
##' "We first make a data.frame called tx2gene with two columns: 1) transcript ID and 2) gene ID.
##' The column names do not matter but this column order must be used. The transcript ID must be
##' the same one used in the abundance files."
# tx2gene_dummy = cbind.data.frame(transcriptID=c('ENST00000631435', 'ENST00000434970', 'ENST00000621592'), geneID=c('TRBD1-201', 'TRDD2-201', 'MYC'))
tx2gene_dummy = cbind.data.frame(transcriptID=transc$tx_id, geneID=ag$gene_name[match(transc$gene_id, ag$gene_id)])
names(files) = make.names(gsub("/abundance.tsv", "", gsub("../Kallisto//", "", (files))))
txi.kallisto <- tximport(files, type = "kallisto", txOut = FALSE, tx2gene = tx2gene_dummy, ignoreTxVersion = TRUE)
head(txi.kallisto$counts)

# Kallisto with with TSV files
#-------------------------
# need tx2gene
# keep code here for future use (kallisto is droppping hdf5)
if(FALSE){
  files <- file.path("Kallisto2", c("JBLAB-19903","JBLAB-19920"), "abundance.tsv")
  names(files) <- c("JBLAB-19903","JBLAB-19920")
  
  txi.kallisto.tsv <- tximport(files, type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE)
  head(txi.kallisto.tsv$counts)
}

if(FALSE){ # feature counts
  # Note: if you are using an Ensembl transcriptome, the easiest way to create the tx2gene data.frame is to use the ensembldb packages. The annotation packages can be found by version number, and use the pattern EnsDb.Hsapiens.vXX. The transcripts function can be used with return.type="DataFrame", in order to obtain something like the df object constructed in the code chunk above. See the ensembldb package vignette for more details.
  
  samples <- read.table(file.path(dir, "samples.txt"), header = TRUE)
  samples
  
  # concat featureCount outcome for each sample into a single table:
  dd <- matFromFeatureCounts(sampleTable = splSht[,c("SampleName","ftcFn")],
                             directory = "samplebam")
}

# merge TPM matrix and gene annotation
#-------------------------
tpmMat <- (txi.kallisto$counts) %>%
  data.frame() %>%
  tibble::rownames_to_column("gene_id") %>%
  dplyr::left_join(data.frame(ag), by="gene_id")

tpmMat <- tpmMat %>% select(colnames(ag), starts_with('JBLAB')) %>%
  select(-seq_coord_system)
tpmMat <- tpmMat %>% mutate_if(is.numeric, round, 1)
#??Write the TPM matrix to file:
tmpFn <- sprintf("%s/TpmMat/%s_tpm.csv", projDir, projName)
# write.csv(tpmMat, file=tmpFn, row.names = FALSE)

##LM: adding gene_name as gene_id
tpmMat$gene_name = tpmMat$gene_id

# correlation gene expression and CNV.
#-------------------------

# format the TPM matrix
#-------------
#??geneDf uses 'sample' eg 119025org, aka 'organoid_name' in splSht
#??TPM:
## I make up a list of genes of interest
# geneDf_dummy = cbind.data.frame(Gene=c('MYC', 'CCNE1', 'PIK3CA', 'TERT', 'KRAS', 'PTEN', 'RB1', 'AKT1',
#                                        'AKT2', 'PARP1', 'PARP2', 'ATM', 'ATR', 'WEE1', 'TOP1', 'TUBB1'))
geneDf_dummy = cbind.data.frame(Gene=ag$gene_name)
genesToKeep <- as.character(unique(geneDf_dummy$Gene))
tpmMatOrig <- tpmMat 

tpmMat <- tpmMat %>% filter(gene_name %in% genesToKeep) %>%
  select(gene_name, colnames((txi.kallisto$counts)))

#??convert TPM matrix to long format to match that keeping CN sent by Maria:
tpmMatLong <- tpmMat %>% tidyr::pivot_longer(-gene_name, names_to = "SampleName", values_to = "tpm")

# merge with CN matrix, using {gene_name, SampleName} IDs.
#-------------
geneDfOrig <- geneDf
geneDfOrig$sample <- as.character(geneDfOrig$sample)

geneDf <- geneDfOrig %>% mutate(organoid_name = sample) %>%
  left_join(splSht[, c("organoid_name", "SampleName")], by="organoid_name")

geneDf2 <- geneDf %>% select(Gene, segVal, SampleName) %>%
  mutate(gene_name = Gene) %>%
  left_join(tpmMatLong, by=c("gene_name", "SampleName"))

#??gene of interest:
#-------------

#??plot single gene
if(FALSE){ # single gene
  
  tmpCn <- geneDf %>% filter(Gene == "BRCA1") %>% pull(segVal)
  
  library(ggplot2)
  tmpdf <- geneDf2 %>% filter(Gene == "BRCA1")
  
  #p <- ggplot(tmpdf, aes(tpm, segVal))
  p <- ggplot(tmpdf, aes(segVal, tpm))
  p + geom_point()
}

# plot all genes of interest at once.
#??so far 42,
#??shown on 6 columns
#??setting the upper bound of the axes to the 90th quantile

# axes limits:
xlim <- c(0, quantile(geneDf2$segVal, probs=0.95))
ylim <- c(0, quantile(geneDf2$tpm, probs=0.95))

# plot:
p <- ggplot(geneDf2, aes(segVal, tpm))
p <- p + geom_point()
p + facet_wrap(~Gene, ncol=6)

p <- p + coord_cartesian(xlim=xlim, ylim=ylim)
p <- p + facet_wrap(~Gene, ncol=6)

# write plot to file
tmpFn <- sprintf("%s/TpmMat/tmpVsSegValInGeneList.png", projDir)
ggsave(tmpFn, plot=p, width = 6*4, height = 7*4, units = "cm")

# write geneDf2 to file
geneDf2 <- geneDf2 %>% mutate_if(is.numeric, round, 1)
tmpFn <- sprintf("%s/TpmMat/tmpVsSegValInGeneList.csv", projDir)
write.csv(geneDf2, file=tmpFn, row.names = FALSE)

#---------------------------------------------------------------------------------------------------#
# ggplot(tpmMatLong, aes(x=SampleName, y=tpm, col=gene_name))+geom_point()

## get the weighted CN state for the genes of interest

ag_subsetchrom = ag[ag$seq_name %in% gsub("chr", "", chromlens$Chrom),]
gr_genes = as(paste0(ag_subsetchrom$seq_name, ':', ag_subsetchrom$gene_seq_start, '-', ag_subsetchrom$gene_seq_end), "GRanges")
gr_genes$gene_name = ag_subsetchrom$gene_name
gr_genes

## only use some genes
gr_genes = gr_genes[gr_genes$gene_name %in% geneDf_dummy$Gene,]

gr_CN0 = as(absCnDf$X, "GRanges")

gr_genes = gr_genes[seqnames(gr_genes) %in% seqnames(gr_CN0),] ## no genes for sex chromosomes

# org_it = 'X119025org'
CN_averages = mclapply(colnames(absCnDf)[-1], function(org_it){
  
  gr_CN = gr_CN0
  values(gr_CN) = absCnDf[,org_it]
  
  ## do per organoid
  gr_CN_org = gr_CN
  seqlevels(gr_CN) ## necessary for mcolAsRleList
  seqlengths(gr_CN) = as.numeric(as.vector(chromlens$Length[match(seqlevels(gr_CN), gsub("chr", "", chromlens$Chrom))])) ## necessary for mcolAsRleList
  seqlevels(gr_genes) = seqlevels(gr_CN) ## necessary for mcolAsRleList
  seqlengths(gr_genes) = seqlengths(gr_CN) ## necessary for mcolAsRleList
  
  gr_genes <- trim(gr_genes)
  gr_CN <- trim(gr_CN)
  
  full_GR <- c(gr_genes, gr_CN_org)
  disjoint_gr <- GenomicRanges::disjoin(full_GR, with.revmap=TRUE, ignore.strand=TRUE)
  
  disjoint_gr_revmap_first = sapply(disjoint_gr$revmap, function(i) i[1])
  
  ## keep only those in which there is an overlap with a gene, i.e. revmap has to include the idxs 1:length(gr_genes)
  disjoint_gr = disjoint_gr[disjoint_gr_revmap_first %in% 1:length(gr_genes),]
  disjoint_gr_revmap_first = sapply(disjoint_gr$revmap, function(i) i[1])
  disjoint_gr_revmap_second = sapply(disjoint_gr$revmap, function(i) i[2])
  table(is.na(disjoint_gr_revmap_second))
  
  ## remove sections which only contain non-genes
  disjoint_gr[is.na(disjoint_gr_revmap_second),]
  
  ## get the idx of the copy number segments
  idx_gr = disjoint_gr_revmap_second-length(gr_genes)
  is.na(idx_gr)
  
  idx_gr[idx_gr < 0 ] = NA
  
  ## there are instances of overlaps between genes
  # dontkeep_disjoint = ( (idx_gr < 0) | is.na(idx_gr) )
  
  # disjoint_gr = disjoint_gr[!dontkeep_disjoint,]
  # disjoint_gr_revmap_first = disjoint_gr_revmap_first[!dontkeep_disjoint]
  # disjoint_gr_revmap_second = disjoint_gr_revmap_second[!dontkeep_disjoint]
  # idx_gr = idx_gr[!dontkeep_disjoint]
  
  disjoint_gr$CN_val = gr_CN_org$X[idx_gr]
  disjoint_gr$CN_val[is.na(disjoint_gr$CN_val)] = 2
  
  disjoint_gr$revmap = disjoint_gr_revmap_first
  seqlevels(disjoint_gr) = as.character(unique(seqnames(disjoint_gr)))
  seqlengths(disjoint_gr) = as.numeric(as.vector(chromlens$Length[match(seqlevels(disjoint_gr), gsub("chr", "", chromlens$Chrom))]))
  seqlevels(gr_genes) = seqlevels(disjoint_gr)
  seqlengths(gr_genes) = seqlengths(disjoint_gr)
  
  CN_RleList = mcolAsRleList(disjoint_gr, "CN_val") ## that takes forever even when length(disjoint_gr) is a VERY low value
  averageCN = binnedAverage(bins = gr_genes, numvar = CN_RleList, varname = "CN_bin_averaged")
  
  return(averageCN)
})
names(CN_averages) = colnames(absCnDf)[-1]

saveRDS(CN_averages, file = paste0("../robjects/CN_averages_", type_counts, "_", gsub("-|:| ", "", Sys.time()), ".RDS"))

averaged_CN_df = lapply(CN_averages, function(i) cbind.data.frame(gene_name=i$gene_name, CN=i$CN_bin_averaged))
names(averaged_CN_df) = names(CN_averages)

## add RNASeq counts

rownames(tpmMat) = tpmMat$gene_name
tpmMat = tpmMat[,-1]

tpmMat_melt = (reshape2::melt(t(tpmMat)))
tpmMat_melt$Var1

## match names
name_correspondence = openxlsx::read.xlsx("/mnt/scratcha/fmlab/morril01/Vias_git/Organoids_Compositional_Analysis/DE_resistant_sensitive/files/PDOnameProperSample_sWGS_RNAseq.xlsx")

tpmMat_melt$Var1 = name_correspondence$ID[match(gsub("[.]", "", tpmMat_melt$Var1), gsub("-", "", name_correspondence$sampleNameRNAseq))]

averaged_CN_df_melt = (reshape2::melt(averaged_CN_df))
averaged_CN_df_melt$L1 = gsub("^X", "", averaged_CN_df_melt$L1)
# tpmMat_melt = tpmMat_melt[match(averaged_CN_df_melt$L1, tpmMat_melt$Var1),]

joint_counts_CN = cbind.data.frame(counts=tpmMat_melt[match(paste0(averaged_CN_df_melt$L1, averaged_CN_df_melt$gene_name),
                                                            paste0(tpmMat_melt$Var1, tpmMat_melt$Var2)),], CN=averaged_CN_df_melt)

joint_counts_CN$counts.value = as.numeric(as.character(joint_counts_CN$counts.value))
saveRDS(joint_counts_CN, file = paste0("../robjects/joint_counts_CN_", type_counts, "_", gsub("-|:| ", "", Sys.time()), ".RDS"))

ggplot(joint_counts_CN, aes(y=(counts.value), x=CN.value, label=CN.L1))+geom_point(aes(col=CN.gene_name))+geom_label_repel()

plot(joint_counts_CN$counts.value,
     joint_counts_CN$CN.value)

plot(log(joint_counts_CN$counts.value),
     log(joint_counts_CN$CN.value))


## example that matching has been done correctly
'JBLAB.19936'
'119025org'
joint_counts_CN[400,]
tpmMat['TRNAU1AP',]
averaged_CN_df$X119025org[averaged_CN_df$X119025org$gene_name == 'TRNAU1AP',]


ggplot(joint_counts_CN %>% filter(CN.L1 == '119025org' ), aes(x=CN.value, y=counts.value))+geom_point()


# ## showing that it works
# c[averageCN$gene_name==gr_genes[13,]$gene_name,]
# disjoint_gr[disjoint_gr_revmap_first == 13,]
# 
# averageCN
# 
# averaged_CN = cbind.data.frame(gene_name=averageCN$gene_name, CN=averageCN$CN_bin_averaged)
# 
# CN_and_counts = cbind.data.frame(averaged_CN[match(tpmMat$gene_name, averaged_CN$gene_name),], tpmMat)

